src/adapter/http/src/smart_protocol/axum_server_pull_protocol.rs

```diff
@@ -45,40 +46,70 @@ impl AxumServerPullProtocolInstance {
     }
 
     pub async fn serve(mut self) {
-        let pull_request = match self.handle_pull_request_initiation().await {
-            Ok(pull_request) => pull_request,
-            Err(e) => {
-                tracing::debug!("Pull process aborted with error: {}", e);
-                return;
+        match self.pull_main_flow().await {
+            Ok(_) => {
+                tracing::debug!("Pull process success");
+            }
+            Err(e) => match &e {
+                PullServerError::Internal(int_err) => {
+                    if let Err(write_err) = axum_write_close_payload::<DatasetPullResponse>(
+                        &mut self.socket,
+                        Err(DatasetPullRequestError::Internal(TransferInternalError {
+                            phase: int_err.phase.clone(),
+                            error_message: "Internal error".to_string(),
+                        })),
+                    )
+                    .await
+                    {
+                        tracing::error!(
+                          error = ?write_err,
+                          error_msg = %write_err,
+                          "Failed to send error to client with error",
+                        );
+                    };
+                    tracing::error!(
+                      error = ?e,
+                      error_msg = %e,
+                      "Push process aborted with internal error",
+                    );
+                }
+                _ => tracing::error!(
+                  error = ?e,
+                  error_msg = %e,
+                  "Push process aborted with error",
+                ),
+            },
+        }
+
+        // After we finish processing to ensure graceful closing
+        // we wait for client acknowledgment. And to handle
+        // custom clients without logic of close acknowledgment
+        // we will give 5 secs timeout
+        let timeout_duration = Duration::from_secs(5);
+        tokio::select! {
+            _ = wait_for_close(&mut self.socket) => {}
+            _ = tokio::time::sleep(timeout_duration) => {
+                tracing::debug!("Timeout reached, closing connection");
             }
         };
```

Comment:
A better way to express this might be via [tokio::time::timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html)